flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala

```diff
@@ -97,6 +97,7 @@ class FlinkPlannerImpl(
       typeFactory,
       SqlValidator.Config.DEFAULT
         .withIdentifierExpansion(true)
+        .withCallRewrite(false) // Disables the rewrite of COALESCE
```

Comment:
Does it _only_ do that? Did you look into whether it has other effects?

Reply:
Seems like the getter for that flag is used only at `SqlOperatorImpl` line 1193, which invokes `SqlOperator#rewriteCall` that is used only by `SqlProcedureCallOperator`, `SqlJdbcFunctionCall`, `SqlCoalesceFunction`, `SqlDatePartFunction`, `SqlNewOperator` and `SqlNullifFunction`. Among all of these, my understanding is that we use on coalesce, sql data part function and null if (which seems to me the same of if_null). And in fact, this change breaks a test here: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=23992&view=ms.vss-test-web.build-test-results-tab&runId=1115000&resultId=100738&paneView=debug

The way to hack around this is to remove this line change and add a condition to the if at `SqlOperatorImpl` line 1193, but I don't really like it :) Do you have any other alternative?

Reply:
Opened an issue in the calcite issue tracker: https://issues.apache.org/jira/browse/CALCITE-4775

Reply:
I've solved this issue disabling COALESCE operator: https://github.com/apache/flink/pull/17256/files#diff-a0641b04c21f8ca7b7b8131e516aa6cddbaf66c6ebd01413f4b67f4d80c778c2L1114

Reply:
This creates another issue though, it disables constant folding: 

```
optimized rel plan expected:<
LogicalProject([$f0=[CAST(10):INTEGER]], empno=[$0], ename...> but was:<
LogicalProject([deptno=[COALESCE(10, $8)]], empno=[$0], ename...>
```

Reply:
the disabling of all rewrite would have been problematic, having different plans for the COALESCE calls sounds ok to me. fyi: you can simply delete the affected xml file and rerun the entire test class, this will regenerate the file.

Reply:
I've added a new commit with the re-generated it test cases and i've removed the RewriteCoalesceRule